### PR TITLE
feat: convert nanoseconds to milliseconds automatically

### LINKED ISSUE #558: Deprecate custom UDF time_bucket in favor of date_bin
### Describe This Problem

In early version of CeresDB, there is a custom UDF `time_bucket` to align time of values, but now datafusion support it out of box, so I suggest deprecate our `time_bucket`, and use `data_bin` instead. Ex:

```sql
SELECT
      DATE_BIN(INTERVAL '15' minute, time, TIMESTAMP '2001-01-01T00:00:00Z') AS time,
      val
    FROM (
      VALUES
        (TIMESTAMP '2021-06-10 17:05:00Z', 0.5),
        (TIMESTAMP '2021-06-10 17:19:10Z', 0.3)
      ) as t (time, val)


        "+---------------------+-----+",
        "| time                | val |",
        "+---------------------+-----+",
        "| 2021-06-10 17:00:00 | 0.5 |",
        "| 2021-06-10 17:15:00 | 0.3 |",
        "+---------------------+-----+",
```

### Proposal

1. Completely remove it before 1.0 is out.
2. If cannot remove it, add a deprecation warning.

### Additional Context

https://github.com/apache/arrow-datafusion/issues/3015

### PR DESCRIPTION
**Pull request by dust1**
## Which issue does this PR close?

Prepare for #558 

## What changes are included in this PR?

I removed the time_bucket function and used datafusion's built-in date_bin function

## Are there any user-facing changes?

The time_bucket function is missing and needs to be replaced with the date_bin function. 
```
SELECT DATE_BIN(INTERVAL '15' minute, `timestamp`, TIMESTAMP '2001-01-01T00:00:00Z')  FROM `02_function_time_bucket_table`;
```
```
{
-"rows": [
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1656777600000000000
},
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1659484800000000000
},
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1659571200000000000
},
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1659576600000000000
},
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1659576600000000000
},
-{
"datebin(IntervalDayTime(\"900000\"),02_function_time_bucket_table.timestamp,Utf8(\"2001-01-01T00:00:00Z\"))": 1659576600000000000
}
]
}
```

## How does this change test

 integration-test case

### CODE REVIEW

**Comment by dust1** [comment]
data_bin returns nanoseconds, whereas Timestamp in ColumnBlock takes milliseconds, and I couldn't find the API for converting datafusion internally, so I handled it separately.
The down_cast of ArrayRef bothered me for a long time, until I realized that it really was simply a conversion of an abstract Array down to a concrete type, not a conversion of a data type. The other day I got into the wrong mindset😌

**Comment by jiacai2050** [comment]
>  and I couldn't find the API for converting datafusion internally, so I handled it separately.

There is a function for this `to_timestamp_millis`.
- https://docs.ceresdb.io/sql/functions/scalar_functions.html#temporal-functions


I see you have deleted `time_bucket`, I'm afraid this is not what the original issue means. 
We plan to reimplement `time_bucket` based on `date_bin`, not remove it, since this would be a breaking change.

**Comment by dust1** [comment]
my mistake, I will recover it

**Comment by dust1** [comment]
I tried the to_timestamp_millis function to convert, but there was an error. 
```
Unsupported data type Timestamp(Nanosecond, None) for function to_timestamp_millis. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```
So I left it the way it was for now.

**Comment by jiacai2050** [comment]
> Unsupported data type Timestamp(Nanosecond, None) for function to_timestamp_millis. 

When will this error arise?

```sql
SELECT
    to_timestamp_millis(date_bin(INTERVAL '15' minute, time, '2001-01-01T00:00:00Z')) AS t,
    count(1)
FROM
    demo
GROUP BY
    t
ORDER BY
    t
```
SQL above can run successfully.

File: integration_tests/cases/common/function/data_bin.sql
Comment (jiacai2050) [inline integration_tests/cases/common/function/data_bin.sql:1]: This file should be `date_bin.sql`

File: integration_tests/cases/common/function/data_bin.result [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    (1659571200000, 4),
+    (1659484800000, 5),
+    (1656777600000, 6);
+
+affected_rows: 6
+
+SELECT DATE_BIN(INTERVAL '30' second, `timestamp`, TIMESTAMP '2001-01-01T00:00:00Z') as `time` FROM `02_function_date_bin_table`;
+
+time,
+Timestamp(1656777600000),
```
Comment (jiacai2050) [inline integration_tests/cases/common/function/data_bin.result:30]: Could you rewrite all those query as
```sql
SELECT 
timestamp,
DATE_BIN(INTERVAL '30' second, `timestamp`, TIMESTAMP '2001-01-01T00:00:00Z') as `time`
FROM `02_function_date_bin_table`
order by timestamp;
```
Your current test case have no specific order, and is hard to tell whether it's right.

**Comment by jiacai2050** [comment]
@dust1  I have left a few comments, have a look when you're free.

Also, I found this PR isn't what the title says, this PR add nanoseconds support, not deprecate old `time_bucket`.

By deprecate `time_bucket`, it's meant to remove old implementation of time_bucket, and reimplement using `date_bin`.

But I think your can do this replace work in next PR, as this PR does preparation for that.

**Comment by jiacai2050** [comment]
@dust1 The CI is broken 😅

**Comment by dust1** [comment]
Oops, an unintended effect😂

**Comment by jiacai2050** [comment]
@dust1 I think this PR is ready for merge, I will push some minor changes, please don't update this branch any more.

**Comment by dust1** [comment]
ok

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**